src/hotspot/share/gc/parallel/psCardTable.hpp

```diff
@@ -34,6 +34,85 @@ class PSPromotionManager;
 
 class PSCardTable: public CardTable {
  private:
+  static constexpr size_t num_cards_in_stripe = 128;
+  static_assert(num_cards_in_stripe >= 1, "progress");
+
+  class StripeShadowTable {
+    CardValue _table[num_cards_in_stripe];
+    const CardValue* _table_base;
+
+   public:
+    StripeShadowTable(PSCardTable* pst, MemRegion stripe) :
+      _table_base(_table - (uintptr_t(stripe.start()) >> _card_shift)) {
+      // Old gen top is not card aligned.
+      size_t copy_length = align_up(stripe.byte_size(), _card_size) >> _card_shift;
+      size_t clear_length = align_down(stripe.byte_size(), _card_size) >> _card_shift;
```

Comment:
Since old-gen-top before scavenging might not be card-aligned, it's unsafe to clear it; hence the conservative (align-down) calculation. However, the right shift will do implicit align-down as well, so it is probably not needed. Better be explicit, I was thinking. Either is fine, I guess.

Reply:
Likely I do not completely understand what you are saying but this would be my explanation why the `align_down` for `clear_length` is needed.
`T := old_gen->object_space()->top()` is not necessarily card aligned at scavenge start. We must not clear the card for `T` if an object was copied there because it was promoted and it has a reference to a young object on that card.

Reply:
```
Example

We cannot clear card n containing old gen top T because we won't scan the promoted
objects on card n and we can only clear cards if we scan all objects on them afterwards.


  card n-1             card n               card n+1
 +--------------------+--------------------+--------------------
 |                    |        .  Promoted |
 |                    |        .  Objects  |
 |                    |        .           |
 +--------------------+--------------------+--------------------
                               ^
                               |
                               T
                      old gen top at scavenge start
                    / end of last stripe
```

Reply:
Okay, I think I understood this now. I would probably just fill up the last card before scavenge, but this explanation makes sense. Please add a comment there for the next person.